File: rust/src/reverse_linked_list.rs
use std::{
    alloc::{Layout, alloc, dealloc},
    ptr::{self, drop_in_place, null_mut, write},
};

pub struct JosieLinkedList<T> {
    list: *mut JosieLinkedListSegment<T>,
    length: usize,
    tail_ptr: *mut JosieLinkedListSegment<T>,
}

impl<T> Drop for JosieLinkedList<T> {
    fn drop(&mut self) {
        unsafe {
            let mut next = self.list;
            while !next.is_null() {
                drop_in_place(&mut (*next).data);
                let to_dealloc = next;
                next = (*next).ptr;
                dealloc(
                    to_dealloc as *mut u8,
                    Layout::new::<JosieLinkedListSegment<T>>(),
                );
            }
        }
    }
}

struct JosieLinkedListSegment<T> {
    ptr: *mut JosieLinkedListSegment<T>,
    data: T,
}

impl<T> JosieLinkedListSegment<T> {
    fn new(element: T) -> *mut Self {
        unsafe {
            let ptr =
                alloc(Layout::new::<JosieLinkedListSegment<T>>()) as *mut JosieLinkedListSegment<T>;
            ptr.is_null().then(|| panic!("Alloc failed"));
            (*ptr).ptr = null_mut();
            ptr::write(&mut (*ptr).data, element);
            ptr
        }
    }
}

impl<T> JosieLinkedList<T> {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn push_back(&mut self, element: T) {
        let segment = JosieLinkedListSegment::new(element);

        self.length += 1;
        if self.list.is_null() {
            self.list = segment;
            self.tail_ptr = segment;
            return;
        }

        unsafe {
            (*self.tail_ptr).ptr = segment;
            self.tail_ptr = segment;
        }
    }

    pub fn push_front(&mut self, element: T) {
        let segment = JosieLinkedListSegment::new(element);
        self.length += 1;
        unsafe {
            (*segment).ptr = self.list;
        }
        self.list = segment;
    }

    pub fn reverse(&mut self) {
        if !self.list.is_null() {
            unsafe {
                let mut original = (*self.list).ptr;
                let mut reversed = self.list;
                (*reversed).ptr = null_mut();
                self.tail_ptr = reversed;
                while !original.is_null() {
                    let next = (*original).ptr;
                    (*original).ptr = reversed;
                    reversed = original;
                    original = next;
                }
                self.list = reversed;
            }
        }
    }

    pub fn iter_mut<'a>(&'a mut self) -> JosieLinkedListIterMut<'a, T> {
        JosieLinkedListIterMut {
            segment: self.list,
            _list: self,
        }
    }

    pub fn iter<'a>(&'a self) -> JosieLinkedListIter<'a, T> {
        JosieLinkedListIter {
            _list: self,
            segment: self.list,
        }
    }

    pub const fn len(&self) -> usize {
        self.length
    }

    pub const fn is_empty(&self) -> bool {
        self.length == 0
    }
}

pub struct JosieLinkedListIterMut<'a, T> {
    _list: &'a mut JosieLinkedList<T>,
    segment: *mut JosieLinkedListSegment<T>,
}

impl<'a, T> Iterator for JosieLinkedListIterMut<'a, T> {
    type Item = &'a mut T;

    fn next(&mut self) -> Option<&'a mut T> {
        if self.segment.is_null() {
            return None;
        }
        unsafe {
            let out = &mut (*self.segment).data;
            self.segment = (*self.segment).ptr;
            Some(out)
        }
    }
}

pub struct JosieLinkedListIter<'a, T> {
    _list: &'a JosieLinkedList<T>,
    segment: *mut JosieLinkedListSegment<T>,
}

impl<'a, T> Iterator for JosieLinkedListIter<'a, T> {
    type Item = &'a T;

    fn next(&mut self) -> Option<&'a T> {
        if self.segment.is_null() {
            return None;
        }
        unsafe {
            let out = &(*self.segment).data;
            self.segment = (*self.segment).ptr;
            Some(out)
        }
    }
}

impl<T> Default for JosieLinkedList<T> {
    fn default() -> Self {
        Self {
            list: ptr::null_mut(),
            length: 0,
            tail_ptr: ptr::null_mut(),
        }
    }
}

File: python/src/reverse_linked_list.py
from typing import TypeVar, Generic

T = TypeVar("T")

#i really dont like the way in which python obscures pointers, this shouldnt work, and yet it does
class JosieLinkedList(Generic[T]):
    def __init__(self):
        self.list:_JosieLinkedListSegment[T] | None = None
        self.tail_ptr:_JosieLinkedListSegment[T] | None = None
        self.length:int = 0

    def push_back(self, element:T):
        segment = _JosieLinkedListSegment[T](element)
        self.length +=1

        if self.list == None:
            self.list = segment
            self.tail_ptr = segment
            return

        assert self.tail_ptr is not None
        self.tail_ptr.ptr = segment
        self.tail_ptr = segment

    def push_front(self, element:T):
        segment = _JosieLinkedListSegment[T](element)
        self.length +=1
        segment.ptr = self.list
        self.list = segment

    def reverse(self):
        if self.list != None:
            original = self.list.ptr
            reversed = self.list
            reversed.ptr = None
            self.tail_ptr = reversed
            while original != None:
                next = original.ptr
                original.ptr = reversed
                reversed = original
                original = next
            self.list = reversed

    def len(self) -> int:
        return self.length

    def is_empty(self) -> bool:
        return self.length == 0

    def __iter__(self):
        return _JosieLinkedListIter[T](self.list)

class _JosieLinkedListSegment(Generic[T]):
    def __init__(self, data:T):
        self.data:T = data
        self.ptr:_JosieLinkedListSegment[T] | None = None

class _JosieLinkedListIter(Generic[T]):
    def __init__(self, segment:_JosieLinkedListSegment[T] | None):
        self.segment = segment

    def __next__(self) -> T:
        if self.segment == None:
            raise StopIteration
        out = self.segment.data
        self.segment = self.segment.ptr
        return out
